mobilecoind-dev-faucet/src/worker.rs

```diff
@@ -312,12 +331,36 @@ impl Worker {
     /// to either successfully send the TxOut and use its oneshot::Sender to
     /// report the result from consensus, or, to drop the oneshot::Sender,
     /// reporting an error using the TxOut.
-    pub fn get_utxo(&self, token_id: TokenId) -> Result<UtxoRecord, String> {
+    pub fn get_utxo(&self, token_id: TokenId) -> Result<UtxoRecord, GetUtxoError> {
         if let Some(receiver) = self.receivers.get(&token_id) {
             receiver.get_utxo()
         } else {
-            Err(format!("Unknown token id: {}", token_id))
+            Err(GetUtxoError::UnknownTokenId)
+        }
+    }
+
+    /// Get any available utxo. This is only used by slam, which just wants
+    /// utxos and doesn't particularly care about token ids.
+    ///
+    /// The error returned is the "least serious" error, i.e. most likely to be
+    /// resolved, that prevents us from getting tokens of one of the token ids.
+    /// This helps slam worker decide whether to try again later or give up.
+    pub fn get_any_utxo(&self) -> Result<UtxoRecord, GetUtxoError> {
+        let mut least_serious_error = None;
+        for (_, receiver) in self.receivers.iter() {
+            match receiver.get_utxo() {
+                Ok(utxo) => return Ok(utxo),
+                Err(err) => {
+                    // Note: this depends on the way rust derives partial ord for enums.
+                    // None is always the least, and then within GetUtxoError, the top-to-bottom
+                    // listing order in code determines the order, with the top being the least
+                    least_serious_error = max(least_serious_error, Some(err));
```

Comment:
yeah the problem I had was this:

* If I get one error "faucet is busy" and two errors "faucet is depleted", I want to return "faucet is busy" so that the caller can know that if they try again later they may get more UTXOs
* So we could have ordered the enum so that it's like "faucet is busy" < "faucet is depleted" < "channel closed", and then take the `min` here.
* However, Rust implements `Ord` for `Option<T>` in such a way that `None` is less than everything else.
* Since the ordering is pretty arbitrary, it seemed like the easiest thing was to spin it around so that the least serious errors compare greater, so that then max will do what I want with `Option<T>`. At least, that is the least code overall.
* Another approach might be like,

```
if let Err(prev_err) = least_serious_error {
    least_serious_error = min(prev_err, err);
} else {
    least_serious_error = Some(err);
}
```